### PR TITLE
implement config json exporter (#575)

### PR DESCRIPTION
**Pull request by zzeekk**
see #575 and usage in sdl-visualizer: https://github.com/smart-data-lake/sdl-visualization/pull/26

### CODE REVIEW

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  def exportConfigJson(config: ConfigJsonExporterConfig): String = {
+    val defaultHadoopConf: Configuration = new Configuration()
+    val sdlConfig = ConfigLoader.loadConfigFromFilesystem(config.configPaths, defaultHadoopConf)
+    // remove additional config paths introduced by system properties...
+    val configKeysToRemove = sdlConfig.root.keySet().asScala.diff(Set(ConfigParser.CONFIG_SECTION_ACTIONS, ConfigParser.CONFIG_SECTION_CONNECTIONS, ConfigParser.CONFIG_SECTION_DATAOBJECTS, ConfigParser.CONFIG_SECTION_GLOBAL))
+    val reducedSdlConfig = configKeysToRemove.foldLeft(sdlConfig)((config,key) => config.withoutPath(key))
+    // enrich origin of first class config objects
+    val descriptionRegex = "(.*): ([0-9]+)(-[0-9]+)?".r
+    val enrichedSdlConfig = reducedSdlConfig.root().keySet().asScala.diff(Set(ConfigParser.CONFIG_SECTION_GLOBAL)).foldLeft(reducedSdlConfig){
```
Comment (Geheiner) [inline sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala]: It's a bit hard to follow. Basically you just want to list a hardcoded list of config sections to enrich, right?
```suggestion
val configSectionsToEnrich =  Set(ConfigParser.CONFIG_SECTION_ACTIONS, ConfigParser.CONFIG_SECTION_CONNECTIONS, ConfigParser.CONFIG_SECTION_DATAOBJECTS)
    val enrichedSdlConfigToExport =configSectionsToEnrich.foldLeft(sdlConfigToExport){
```

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      .text("Whether to add an additional property 'origin' including source filename and line number to first class configuration objects.")
+    help("help").text("Display the help text.")
+  }
+
+  /**
+   * Takes as input a SDL Config and exports it as one json document, everything resolved.
+   * Additionally a separate file with the mapping of first class config objects to source code origin is created.
+   */
+  def main(args: Array[String]): Unit = {
+    val config = ConfigJsonExporterConfig()
```
Comment (Geheiner) [inline sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala]: rename to exporterConfig everywhere. We have enough configs lying around :D

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        // write file
+        logger.info(s"Writing config json to file ${config.filename}")
+        Files.write(Paths.get(config.filename), configAsJson.getBytes, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)
+
+      case None =>
+        logAndThrowException(s"Aborting ${appType} after error", new ConfigurationException("Couldn't set command line parameters correctly."))
+    }
+  }
+
+  def exportConfigJson(config: ConfigJsonExporterConfig): String = {
```
Comment (Geheiner) [inline sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala]:
```suggestion
  def exportConfigJson(exporterConfig: ConfigJsonExporterConfig): String = {
```

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    val defaultHadoopConf: Configuration = new Configuration()
+    val sdlConfig = ConfigLoader.loadConfigFromFilesystem(config.configPaths, defaultHadoopConf)
+    // remove additional config paths introduced by system properties...
+    val configKeysToRemove = sdlConfig.root.keySet().asScala.diff(Set(ConfigParser.CONFIG_SECTION_ACTIONS, ConfigParser.CONFIG_SECTION_CONNECTIONS, ConfigParser.CONFIG_SECTION_DATAOBJECTS, ConfigParser.CONFIG_SECTION_GLOBAL))
+    val reducedSdlConfig = configKeysToRemove.foldLeft(sdlConfig)((config,key) => config.withoutPath(key))
```
Comment (Geheiner) [inline sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala:62]: sdlConfigToExport?
```suggestion
    val sdlConfigToExport= configKeysToRemove.foldLeft(sdlConfig)((config,key) => config.withoutPath(key))
```
The
(if (config.enrichOrigin) enrichedSdlConfig else reducedSdlConfig
confused me a bit with that naming

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            val objectConfig = config.getConfig(s"$sectionKey.$objectKey")
+            // parse origins description, as we can not access the detailed private properties
+            val (path, lineNumber, endLineNumber) = objectConfig.origin.description match {
+              case descriptionRegex(path, lineNumber, endLineNumber) =>
+                (path, lineNumber.toInt, Option(endLineNumber).map(_.toInt))
+            }
+            val config1 = config
+              .withValue(s"$sectionKey.$objectKey.origin.path", ConfigValueFactory.fromAnyRef(path))
+              .withValue(s"$sectionKey.$objectKey.origin.lineNumber", ConfigValueFactory.fromAnyRef(lineNumber))
+            if (endLineNumber.isDefined) config1.withValue(s"$sectionKey.$objectKey.origin.endLineNumber", ConfigValueFactory.fromAnyRef(endLineNumber.get))
```
Comment (Geheiner) [inline sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala:80]: FYI, endLineNumber is never included in the test file produced by the unit test

**Review by Geheiner — DISMISSED** [review]
Looks good, I would just rename some variables and make sure endLineNumber is working

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala (hunk 1 above)
Comment (zzeekk) [reply sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala]: Good point, will do.

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala (hunk 4 above)
Comment (zzeekk) [reply sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala:62]: ok, splitted the "if (config.enrichOrigin) enrichedSdlConfig else reducedSdlConfig" line

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala (hunk 5 above)
Comment (zzeekk) [reply sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala:80]: True, the Hocon parser foresees this format, but doesnt fill endLineNumber for now...
I added a comment in the code.

**Review by Geheiner — APPROVED** [review]
LGTM

File: sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala [hunk 6]
```diff
@@ -37,37 +37,39 @@ object ConfigJsonExporter extends SmartDataLakeLogger {
...
     val descriptionRegex = "(.*): ([0-9]+)(-[0-9]+)?".r
-    val enrichedSdlConfig = reducedSdlConfig.root().keySet().asScala.diff(Set(ConfigParser.CONFIG_SECTION_GLOBAL)).foldLeft(reducedSdlConfig){
+    val configSectionsToEnrich =  Set(ConfigParser.CONFIG_SECTION_ACTIONS, ConfigParser.CONFIG_SECTION_CONNECTIONS, ConfigParser.CONFIG_SECTION_DATAOBJECTS)
+    val enrichedSdlConfig = configSectionsToEnrich.foldLeft(reducedSdlConfig){
       case (config,sectionKey) =>
         config.getConfig(sectionKey).root.keySet().asScala.foldLeft(config) {
           case (config,objectKey) =>
             val objectConfig = config.getConfig(s"$sectionKey.$objectKey")
             // parse origins description, as we can not access the detailed private properties
+            // note that currently endLineNumber is not filled by Hocon parser, even though it is foreseen in the code.
```
Comment (Geheiner) [inline sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala:72]: Just wondering: how will it be filled by hocon parser in the future? Will we do a library update?

Comment (zzeekk) [reply sdl-lang/src/main/scala/io/smartdatalake/meta/configexporter/ConfigJsonExporter.scala:72]: Man interfaces are private. To replace it with our own implementation we could create classes with Hocon package name in our project. Making a PR for Hocon would be cleaner.
But its also not an easy fix...
